### PR TITLE
Set up logging

### PR DESCRIPTION
**Pull request by benjaminjkraft**
This sets us up to use django/python logging facilities.  In particular, we'll now log to a file in addition to serverlog, and error logs that don't raise an error will also go to serverlog and sentry (if configured).  This should make it a lot easier to do things like deprecations, since we'll be able to actually see whether any live site raised a deprecation warning, rather than having to guess based on whether anyone noticed one on a dev server; it will also mean that if we want to make sure we notice if something goes wrong, but still do something reasonable, we can log an error without 500ing.  I also cleaned up our error handling middleware while I was at it.

~~This still needs a bit more testing on a prod-like setup, but I figured I'd let people take a look in the meantime.~~

This is a start on some of the various things suggested in #1237.

### CODE REVIEW

**Comment by luac** [comment]
What's the rationale for having scripts behave differently? It seems like it makes things more complicated.

**Comment by luac** [comment]
Ah, I guess you need it so that scripts log to the console even with DEBUG=False, and if you have that you might as well tailor it a bit more anyway.

**Comment by mgersh** [comment]
It seems like we really have three main cases:
- Code running unattended in prod. This includes the main site, but also includes `dbmail_cron` and `mailgate`. I don't have complaints about the settings for this one, but I'd say that the mail system stuff should be changed to go in this category. If it breaks, we want to get emails.
- The dev server equivalent of the above. No complaints about this one.
- People running scripts by hand or messing around in the shell. Would this log things like somebody making a typo in the shell? That seems likely to make the log full of things that look like problems but that we have no reason to ever care about. Maybe scripts should go to a separate log file? Also, how does this change what I see in the console/shell, if at all?

**Comment by benjaminjkraft** [comment]
Hmm, good point @mgersh, `dbmail_cron` and `mailgate` should probably act like prod servers rather than scripts.  (Their stdout does already go to serverlog, but I doubt anyone ever reads it, so putting errors there (as we currently do) is not the most useful.  I think `dbmail_cron` at least already manually sends an email if it fails in certain ways, but that's also not the most ideal.)

For the shell, I agree it's a bit weird -- as-is it will indeed log errors in the shell.  It's easy enough to turn that off; I guess it probably isn't that useful.  Or we could put it in a separate file, although that's another set of local config to maintain.  I do want to make sure `manage,py` commands still get logged, though, since failures there are no more likely to be user error than failures in prod scripts, and errors are only a little more likely to be seen by the person running them (since many put out a bunch of spam); maybe those should log to a file but scripts shouldn't.

It will somewhat change what you see in the shell, in that things that were previously `print` statements that are now `logger.info` will be prefixed by `INFO:`.  (We can change that easily, but I think it's more useful than spammy, because they might be `ERROR:` instead of `INFO:`.)  You can also still `print` as normal, it will just be discouraged (forbidden?) in code not intended to be run in the shell.

File: esp/esp/dbmail/models.py [hunk 1]
```diff
@@ -345,7 +348,7 @@ def process(self, debug=False):
                 # all recipients the first time.
                 newtxt, created = TextOfEmail.objects.get_or_create(**newtxt)
                 if not created:
-                    if debug: print 'Skipped duplicate creation of message to %s for message request %d: %s' % (send_to, self.id, self.subject)
+                    logger.warning('Skipped duplicate creation of message to %s for message request %d: %s', send_to, self.id, self.subject)
```
Comment (mgersh) [inline esp/esp/dbmail/models.py]: Will this still error if there's unicode in the user's name?

Comment (benjaminjkraft) [inline esp/esp/dbmail/models.py]: It looks like it prints a unicode error, but continues on, which is also not the most ideal.  In any case I can just make the actual fix while I'm at it.

Comment (benjaminjkraft) [inline esp/esp/dbmail/models.py]: (Going to punt that since it seems like something more complicated is going on, see #1982.)

**Comment by benjaminjkraft** [comment]
Updated the code and did some more testing, this is ready to go as far as I'm concerned.  I believe I've addressed all the comments except the ones I said I wouldn't.

**Comment by mgersh** [comment]
:rocket:

**Pull request merged**